### PR TITLE
Add support for `io.jsonwebtoken:jjwt-*:0.11.5`

### PR DESCRIPTION
**Pull request by linghengqian**
## What does this PR do?

- Fixes https://github.com/oracle/graalvm-reachability-metadata/issues/115 .
- Add support for `io.jsonwebtoken:jjwt-jackson:0.11.5`.
- Add support for `io.jsonwebtoken:jjwt-gson:0.11.5`.
- Add support for `io.jsonwebtoken:jjwt-orgjson:0.11.5`.

## Checklist before merging
- [x] I have properly formatted metadata files (see [CONTRIBUTING](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md) document)
- [x] I have added thorough tests. (see [this](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#Tests))

### CODE REVIEW

File: tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/src/test/java/io_jsonwebtoken/jjwt_impl/Jjwt_implTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import io.jsonwebtoken.io.Encoders;
+import io.jsonwebtoken.security.Keys;
+import org.junit.jupiter.api.Test;
+
+import javax.crypto.SecretKey;
+import java.security.KeyPair;
+import java.util.Date;
+import java.util.UUID;
+
+class Jjwt_implTest {
```
Comment (linghengqian) [inline tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/src/test/java/io_jsonwebtoken/jjwt_impl/Jjwt_implTest.java:23]: Should I say that after executing `./gradlew scaffold --coordinates io.jsonwebtoken:jjwt-impl:0.11.5`, is it a strange thing to use underscore nomenclature to generate class names? Do i need to fix it?

File: tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/build.gradle [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = TestUtils.testedLibraryVersion
+
+dependencies {
+    testImplementation "io.jsonwebtoken:jjwt-api:$libraryVersion"
+    testImplementation "io.jsonwebtoken:jjwt-impl:$libraryVersion"
+    testImplementation "io.jsonwebtoken:jjwt-jackson:$libraryVersion"
```
Comment (linghengqian) [inline tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/build.gradle:19]: - Tests that add GraalVM reachability metadata for `io.jsonwebtoken:jjwt-impl` must add `io.jsonwebtoken:jjwt-api` and `io.jsonwebtoken:jjwt-jackson`. 
- The other two dependencies are not convenient for unit testing. Should I completely separate them into three directories for separate unit testing?

File: metadata/io.jsonwebtoken/jjwt-impl/0.11.5/reflect-config.json [hunk 3]
```diff
@@ -0,0 +1,310 @@
+[
+  {
+    "condition": {
+      "typeReachable": "io.jsonwebtoken.jackson.io.JacksonSerializer"
```
[the remark this answers is not in the document]
Comment (lhazlewood) [reply metadata/io.jsonwebtoken/jjwt-impl/0.11.5/reflect-config.json]: @linghengqian JJWT will try to find 3 implementions: Jackson, Gson, and org.json.  I hope that helps!

[the remark this answers is not in the document]
Comment (linghengqian) [reply metadata/io.jsonwebtoken/jjwt-impl/0.11.5/reflect-config.json]: @lhazlewood Thanks for the reminder! I have completed the verification at https://github.com/linghengqian/graalvm-trace-metadata-smoketest, the GraalVM reachability metadata of jackson and orgjson are generated normally, but the GSON test phase fails, Gson does not support `io.jsonwebtoken.JwtParserBuilder#requireExpiration()`, really surprised me. Is this a unique design of `jjwt-gson`?
- image
- In https://github.com/linghengqian/graalvm-trace-metadata-smoketest only need the following command to verify.
```shell
cd ./jjwt-gson/
./gradlew clean test
```

File: tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/build.gradle (hunk 2 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/build.gradle]: - Solved. Since I can't imagine using `io.jsonwebtoken:jjwt-api:0.11.5` alone, I don't see any need to separate its GraalVM reachability metadata.

File: metadata/io.jsonwebtoken/jjwt-impl/0.11.5/reflect-config.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (linghengqian) [reply metadata/io.jsonwebtoken/jjwt-impl/0.11.5/reflect-config.json]: - I'll tag this question in a new code block.

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                .compact();
+        JwtParserBuilder jwtParserBuilder = Jwts.parserBuilder().setAllowedClockSkewSeconds(3 * 60).setSigningKey(firstKey);
+        assert jwtParserBuilder.build().parseClaimsJws(firstCompactJws).getBody().getSubject().equals("Joe");
+        jwtParserBuilder.requireSubject("Joe").build().parseClaimsJws(firstCompactJws);
+        jwtParserBuilder.requireIssuer("Aaron").build().parseClaimsJws(firstCompactJws);
+        jwtParserBuilder.requireAudience("Abel").build().parseClaimsJws(firstCompactJws);
+        // TODO This is an error caused by GSON's own parsing
+        // jwtParserBuilder.requireExpiration(secondDate).build().parseClaimsJws(firstCompactJws);
+        // jwtParserBuilder.requireNotBefore(firstDate).build().parseClaimsJws(firstCompactJws);
+        // jwtParserBuilder.requireIssuedAt(firstDate).build().parseClaimsJws(firstCompactJws);
```
Comment (linghengqian) [inline tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java:67]: @lhazlewood This problem really made me laugh a little bit. Did Gson drop support for `java.util.Date`? Is it possible for jjwt to handle it additionally?

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/build.gradle [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+String libraryVersion = TestUtils.testedLibraryVersion
+
+dependencies {
+    testImplementation "io.jsonwebtoken:jjwt-gson:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+    testImplementation "io.jsonwebtoken:jjwt-api:$libraryVersion"
+    testImplementation "io.jsonwebtoken:jjwt-impl:$libraryVersion"
+}
```
Comment (linghengqian) [inline tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/build.gradle:21]: - Do I need to bring the `user-code-filter.json` that I personally use to collect GraalVM accessibility metadata into this PR?

Comment (mhalbritter) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/build.gradle:21]: Yes, please do.

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.stream.Stream;
+
+/**
+ * @author linghengqian
+ */
+class Jjwt_gsonTest {
+    @Test
+    void testSignedJWTs() {
+        SecretKey firstKey = Keys.secretKeyFor(SignatureAlgorithm.HS256);
+        assert Jwts.parserBuilder().setSigningKey(firstKey).build().parseClaimsJws(
```
Comment (mhalbritter) [inline tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java:30]: Please use `assertJ` like the other tests do instead of plain java asserts.

Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java]: - I finished.

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/build.gradle (hunk 5 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/build.gradle:21]: - I have included this part.

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java (hunk 4 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java]: - I think this is a separate issue, I will add TODO first.

File: tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/src/test/java/io_jsonwebtoken/jjwt_impl/Jjwt_implTest.java (hunk 1 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-impl/0.11.5/src/test/java/io_jsonwebtoken/jjwt_impl/Jjwt_implTest.java:26]: - I'm not quite sure what the final format should be. I expect that after the modification for the scaffold task is in place, I will modify the class name.

File: tests/src/io.jsonwebtoken/jjwt-jackson/0.11.5/user-code-filter.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+      "excludeClasses": "**"
+    },
+    {
+      "includeClasses": "io.jsonwebtoken.jackson.**"
+    },
+    {
+      "includeClasses": "io.jsonwebtoken.**"
+    }
+  ]
+}
```
Comment (linghengqian) [inline tests/src/io.jsonwebtoken/jjwt-jackson/0.11.5/user-code-filter.json:13]: - In fact, `io.jsonwebtoken:jjwt-jackson`, `io.jsonwebtoken:jjwt-orgjson` and `io.jsonwebtoken:jjwt-gson`, they all require `io.jsonwebtoken:jjwt-api` and `io.jsonwebtoken:jjwt-impl`s full or partial GraalVM reachability metadata. 
- I wonder if there is any easier way to directly import the GraalVM reachability metadata of `io.jsonwebtoken:jjwt-impl` to guarantee `./gradlew clean test -Pcoordinates=io.jsonwebtoken:jjwt-jackson:0.11.5 `, won't the execution fail?

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java (hunk 6 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java]: - I left out part of the unit test unmodified, which I'll deal with later in this PR.

File: tests/src/io.jsonwebtoken/jjwt-jackson/0.11.5/user-code-filter.json (hunk 7 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-jackson/0.11.5/user-code-filter.json:13]: - I don't think GraalVM reachability metadata for `io.jsonwebtoken:jjwt-impl:0.11.5` should be independent in this case.

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java (hunk 6 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java]: - done.

File: tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore [hunk 8]
```diff
@@ -0,0 +1,4 @@
+gradlew.bat
```
Comment (gradinac) [inline tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore:1]: Is this file needed even though we have the top level `.gitignore`? Same goes for the other 2 `.gitignore` files under `jjwt-jackson` and `jjwt-gson`

File: tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class Jjwt_gsonTest {
+    @Test
+    void testSignedJWTs() {
+        SecretKey firstKey = Keys.secretKeyFor(SignatureAlgorithm.HS256);
```
Comment (gradinac) [inline tests/src/io.jsonwebtoken/jjwt-gson/0.11.5/src/test/java/io_jsonwebtoken/jjwt_gson/Jjwt_gsonTest.java:27]: These and the other 2 tests seem to be identical, with the only change being which JSON library dependency we're importing in `build.gradle`. Would it be possible to somehow share these sources between the 3 tests? I'm not sure how exactly to do it though :/ cc @melix @dnestoro

File: metadata/io.jsonwebtoken/jjwt-gson/0.11.5/serialization-config.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (gradinac) [inline metadata/io.jsonwebtoken/jjwt-gson/0.11.5/serialization-config.json:1]: I think we can remove the empty metadata files and their entries from `index.json`

File: tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore (hunk 8 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore:1]: - This is generated in one step by `./gradlew scaffold --coordinates io.jsonwebtoken:jjwt-jackson:0.11.5`. Need to fix https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md?

File: metadata/io.jsonwebtoken/jjwt-gson/0.11.5/serialization-config.json (hunk 10 above)
Comment (linghengqian) [reply metadata/io.jsonwebtoken/jjwt-gson/0.11.5/serialization-config.json]: - Done.

File: tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore (hunk 8 above)
Comment (linghengqian) [reply tests/src/io.jsonwebtoken/jjwt-orgjson/0.11.5/.gitignore:1]: - In fact, once this file is removed, other people running nativeTest will pass unnecessary files to git. It's not like this PR should solve the problem, I marked it as solved.

**Review by gradinac — APPROVED** [review]

**Review by vjovanov — APPROVED** [review]
